rfd/0010-api.md

```diff
@@ -0,0 +1,238 @@
+
+**Go native user experience**
+
+Achieve the user experience in Client with a simplified client constructor.
+This should also transparently handle dialing via a proxy.
+
+```go
+import (
+   "github.com/gravitational/teleport/api/client"
+)
+
+func main() {
+   // TLS client is the only TLS client supported in teleport
+   client, err := client.New(client.Config{
+     Addrs: []string{"localhost:3025"},
+     // ContextDialer is optional context dialer
+     ContextDialer: net.DialContext,
+     // direct TLS credentials config
+     Credentials: client.TLSCreds(tls.Config)
+   })
+   ...
+   defer client.Close()
+
+   ping, err := client.Ping()
+   ...
+   fmt.Printf("Ping: %v\n", ping.Version)
+}
+```
+
+```bash
+$ go mod init
+$ go run main.go
+```
+
+**Credential providers**
+
+Support multiple credential providers with chaining
+
+* client.TLSCreds(*tls.Config) loads creds from TLS config
+* client.PathCreds("path") loads mounted creds from path, detects reloads
+and updates the grpc transport
+
+Support convenience method that supports and understands `~/.tsh/profile`,
+so users can try testing using credentials they got from `tsh login`
+
+```go
+// for testing, use client spec that loads dialer, and credentials from profile
+cfg, err := client.LoadTSHConfig()
+// cfg.ContextDialer sets up dialer using tunnel dialer in tctl
+// TLS creds are loaded from key store, similar how tctl does it in 5.0
+client, err := client.New(cfg)
+```
+
+```bash
+$ tsh login
+# try client
+$ go run main.go
+```
+
+**Auto version checks**
+
+Client should verify that its implementation is compatible with
+Teleport's server version by calling `Ping` and checking the version
+similar to tsh logic.
+
+## Move Access requests code into client
+
+Move `access` workflows code into `client` subpackage with new name `workflows`:
+
+https://github.com/gravitational/teleport-plugins/tree/master/access
+
+Instead of a separate client, `workflows` becomes a subpackage of the api client:
+
+```go
+import (
+   "github.com/gravitational/teleport/api/client"
+   "github.com/gravitational/teleport/api/client/workflows"
+)
+
+func main() {
+   client, err := client.New(...)
+   watcher, err := workflows.NewWatcher(ctx, client, workflows.Watch{State: ...})
+   ...
+   defer watcher.Close()
+
+```
+
+## Provide simplified high level access hooks
+
+Access hooks are too low level, add simple handler similar to http server design
+with a router:
+
+```go
+
+func myTeamRequest(ctx context.Context, req access.Request) (access.Request, error) {
+  return req, trace.AccessDenied("access denied")
+}
+
+
+func main() {
+router := access.Router()
+  // handle access requests from my team only
+  router.HandleFunc(router.Match{Traits: Trait("team", "myteam")}
+  srv := access.Server{
+     Client: clt,
+     Handler: router,
+  }
+```
+
+
+## Improve API documentation user experience
+
+Integrate Swagger API generated code into `mkdocs` using mkdocs plugin.
```

Comment:
Link to what `mkdocs` is.
Also, this doc uses `protoc-gen-doc`, not swagger. Make sure that works with `mkdocs`